junit-platform-engine/src/main/java/org/junit/platform/engine/support/hierarchical/NodeTreeWalker.java

```diff
@@ -25,16 +29,20 @@
 
 	private final LockManager lockManager = new LockManager();
 
-	NodeExecutionAdvisor walk(TestDescriptor testDescriptor) {
+	NodeExecutionAdvisor walk(TestDescriptor rootDescriptor) {
 		NodeExecutionAdvisor advisor = new NodeExecutionAdvisor();
-		walk(testDescriptor, advisor);
+		Preconditions.condition(getExclusiveResources(rootDescriptor).isEmpty(),
+			"Engine descriptor must not declare exclusive resources");
+		rootDescriptor.getChildren().forEach(child -> walk(child, child, advisor));
 		return advisor;
 	}
 
-	private void walk(TestDescriptor testDescriptor, NodeExecutionAdvisor advisor) {
+	private void walk(TestDescriptor globalLockDescriptor, TestDescriptor testDescriptor,
+			NodeExecutionAdvisor advisor) {
 		Set<ExclusiveResource> exclusiveResources = getExclusiveResources(testDescriptor);
 		if (exclusiveResources.isEmpty()) {
-			testDescriptor.getChildren().forEach(child -> walk(child, advisor));
+			advisor.useResourceLock(testDescriptor, lockManager.getLockForResources(singleton(GLOBAL_READ)));
```

Comment:
You could create a constant for the `singleton` locks, no real need to create new instances, same for the read_write below

Reply:
Good catch. I added a `getLockForResource` method that takes a single `ExclusiveResource` now.